File: regression/gold/philox.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NUM_ROUNDS 7
#define MULT_0 0xD2511F53
#define MULT_1 0xCD9E8D57
#define WEYL_0 0x9E3779B9
#define WEYL_1 0xBB67AE85

void mulhilo(uint32_t a, uint32_t b, uint32_t *hi, uint32_t *lo) {
    uint64_t mul = ((uint64_t) a) * ((uint64_t) b);
    *lo = (uint32_t)mul;
    *hi = (uint32_t)(mul >> 32);
}

void single_round(uint32_t ctr[4], const uint32_t key[2]) {
    uint32_t lo0, hi0, lo1, hi1;
    mulhilo(ctr[0], MULT_0, &hi0, &lo0);
    mulhilo(ctr[2], MULT_1, &hi1, &lo1);
    ctr[0] = hi1 ^ key[0] ^ ctr[1];
    ctr[1] = lo1;
    ctr[2] = hi0 ^ key[1] ^ ctr[3];
    ctr[3] = lo0;
}

void philox_compute(const uint32_t ctr[4], const uint32_t key[2], uint32_t out[4]) {
    uint32_t ctr_[4], key_[2];
    memcpy(ctr_, ctr, sizeof(ctr_));
    memcpy(key_, key, sizeof(key_));

    for (size_t i = 0; i < NUM_ROUNDS - 1; ++i) {
        single_round(ctr_, key_);
        key_[0] += WEYL_0; key_[1] += WEYL_1;
    }
    single_round(ctr_, key_);

    memcpy(out, ctr_, sizeof(ctr_));
}

void incr(uint32_t ctr[4]) {
    // series of checking overflow
    if (++ctr[0]) return;
    if (++ctr[1]) return;
    if (++ctr[2]) return;
    ++ctr[3];
}

uint32_t philox(uint32_t ctr[4], const uint32_t key[2], unsigned *selector) {
    uint32_t out[4];
    philox_compute(ctr, key, out);
    incr(ctr);
    uint32_t res = out[*selector];
    *selector = (*selector + 1) & 3;
    return res;
}

int main(int argc, char *argv[]) {
    if (argc != 7) {
        printf("not enough arguments\n");
        return 1;
    }
    uint32_t key[2];
    uint32_t ctr[4];
    key[0] = strtoul(argv[1], NULL, 10);
    key[1] = strtoul(argv[2], NULL, 10);
    ctr[0] = strtoul(argv[3], NULL, 10);
    ctr[1] = strtoul(argv[4], NULL, 10);
    ctr[2] = strtoul(argv[5], NULL, 10);
    ctr[3] = strtoul(argv[6], NULL, 10);
    unsigned selector = 0;

    for (unsigned i = 0; i < 10; ++i) {
        printf("%u\n", philox(ctr, key, &selector));
    }

    return 0;
}

File: regression/Philox.scala
import argon._
import spatial.metadata.memory._

import spatial.dsl._

// --merge=false pir (bin/test)
class Philox_0 extends Philox
class Philox_2 extends Philox(op=2)
class Philox_4 extends Philox(op=4)
class Philox_8 extends Philox(op=8)

@forge.tags.stateful object PhiloxCompute {

    val mult_0:UInt32 = U32(0xD2511F53)
    val mult_1:UInt32 = U32(0xCD9E8D57)
    val weyl_0:UInt32 = U32(0x9E3779B9)
    val weyl_1:UInt32 = U32(0xBB67AE85)

    def mul_32(x: UInt32, y: UInt32): (UInt32, UInt32) = {
        (mulh(x, y), x * y)
    }

    def philox_round(key_0: UInt32,
                     key_1: UInt32,
                     r:     scala.Int,
                     ctr_0: UInt32,
                     ctr_1: UInt32,
                     ctr_2: UInt32,
                     ctr_3: UInt32): (UInt32, UInt32, UInt32, UInt32) = {

        // this could be one hardware block instead of two
        val _key_0:UInt32 = key_0 + (weyl_0 * r.to[UInt32])
        val _key_1:UInt32 = key_1 + (weyl_1 * r.to[UInt32])

        val (hi_0, lo_0) = mul_32(ctr_0, mult_0)
        val (hi_1, lo_1) = mul_32(ctr_2, mult_1)

        val _ctr_0_next = hi_1 ^ _key_0 ^ ctr_1
        val _ctr_1_next = lo_1
        val _ctr_2_next = hi_0 ^ _key_1 ^ ctr_3
        val _ctr_3_next = lo_0

        (_ctr_0_next, _ctr_1_next, _ctr_2_next, _ctr_3_next)
    }

    // hybird between fully spatial and fully temporal
    // support arbirary num_rounds, might not be divisible
    def philox_compute(key_0: UInt32,
                     key_1: UInt32,
                     r:     scala.Int,
                     ctr_0: UInt32,
                     ctr_1: UInt32,
                     ctr_2: UInt32,
                     ctr_3: UInt32,
                     num_rounds: scala.Int): (UInt32, UInt32, UInt32, UInt32) = {
        if (r == num_rounds) {
            (ctr_0, ctr_1, ctr_2, ctr_3)
        } else {
            val (ctr_0_next, ctr_1_next, ctr_2_next, ctr_3_next) = philox_round(key_0, key_1, r, ctr_0, ctr_1, ctr_2, ctr_3)
            philox_compute(key_0, key_1, r+1, ctr_0_next, ctr_1_next, ctr_2_next, ctr_3_next, num_rounds)
        }
    }
}

@spatial abstract class Philox(
    len:scala.Int = 131072,
    tileSize:scala.Int = 8192,
    num_rounds:scala.Int = 7,
    op:scala.Int = 1
) extends SpatialTest { self =>

    val ip:scala.Int = 16

    def philox(ctr_0: UInt32,
               ctr_1: UInt32,
               ctr_2: UInt32,
               ctr_3: UInt32,
               key_0: UInt32,
               key_1: UInt32): Array[UInt32] = {

        val rand_d = DRAM[UInt32](len)

        Accel {
            // Write meta loop for 2^32 blocks, so overflow is easy to handle
            Foreach(len by tileSize par op){tile =>

                val rand_out_fifo = FIFO[UInt32](16)

                Pipe.Foreach(tileSize by 1 par ip){i => 
                    val offset:UInt32 = tile.to[UInt32] + i.to[UInt32]

                    val (ctr_0_next, ctr_1_next, ctr_2_next, ctr_3_next) = PhiloxCompute.philox_compute(key_0, key_1, 0, ctr_0 + offset, ctr_1, ctr_2, ctr_3, num_rounds)

                    // set the ith rand to be ctr[offset % 4]
                    val selector_0:UInt32 = offset & 1.to[UInt32]
                    val selector_1:UInt32 = offset & 2.to[UInt32]
                    val rand_num = mux(selector_0 == 0,
                                        mux(selector_1 == 0, ctr_0_next, ctr_2_next),
                                        mux(selector_1 == 0, ctr_1_next, ctr_3_next))
                    rand_out_fifo.enq(rand_num, true)
                }

                rand_d(tile::tile+tileSize par ip) store rand_out_fifo
            }
        }

        getMem(rand_d)
    }

    def main(args: Array[String]): Unit = {
        val gold = loadCSV1D[UInt32](sys.env("SPATIAL_HOME") + "/test-data/philox_test/rand.csv", "\n")
        val rands = philox(0, 0, 0, 0, 0, 0)

        writeCSV1D(rands, sys.env("SPATIAL_HOME") + "/test-data/philox_test/rand_out.csv", "\n")
        assert(rands == gold)
    }

}

